File: MPW/3.3.1/Examples/CExamples/EditCdev.r
/*------------------------------------------------------------------------------
#
#	Macintosh Developer Technical Support
#
#	EditText Sample Control Panel Device
#
#	EditCdev
#
#	EditCdev.make	-	Make Source
#
#	Copyright © 1988 Apple Computer, Inc.
#	All rights reserved.
#
#	Versions:	1.0					8/88
#
#	Components:	EditCdev.c			August 1, 1988
#				EditCdev.r			August 1, 1988
#				EditCdev.make		August 1, 1988
#
#	EditCdev is a sample Control Panel device (cdev) that 
#	demonstrates the usage of the edit-related messages.  
#	EditCdev demonstrates how to implement an editText item
#	in a Control Panel Device.  It utilizes the new undo, cut, copy,
#	paste, and delete messages that are sent to cdevs in
#	response to user menu selections.
#
#	It is comprised of two editText items that can be edited 
#	and moved between via the mouse or tab key.
#
------------------------------------------------------------------------------*/


#include "Types.r"
#include "SysTypes.r"

type 'hack' as 'STR ';

resource 'vers' (1) {
	0x01, 0x00, release, 0x00,
	verUS,
	"1.00",
	"1.00, Copyright © 1988 Apple Computer, Inc."
};

type 'nrct' {
	integer = $$CountOf(RectArray);
	array RectArray { rect; };
};

type 'mach' {
	unsigned hex integer;	/* Softmask */
	unsigned hex integer;	/* Hardmask */
};

resource 'hack' (0, purgeable) {
	"Control Panel Device, INIT and CODE by Macintosh Developer Technical Support"
};

resource 'BNDL' (-4064, purgeable) {
	'hack', 0,
	{	'ICN#', {0, -4064},
		'FREF', {0, -4064}
	}
};

resource 'ICN#' (-4064, purgeable) {
	{ /* array: 2 elements */
		/* [1] */
		$"00 00 00 00 00 00 3F E0 00 00 48 10 00 00 44 10"
		$"00 01 83 10 00 02 01 90 00 04 01 90 00 04 01 90"
		$"00 04 01 90 00 04 03 90 00 04 02 90 00 04 02 90"
		$"00 08 02 90 00 08 02 90 00 10 06 90 00 20 0C 90"
		$"03 C0 18 90 04 00 3F 10 0F FF C0 10 08 00 00 10"
		$"08 00 00 10 08 00 00 10 08 00 FF 10 08 00 00 10"
		$"08 00 00 10 08 00 00 10 08 00 00 10 07 FF FF E0"
		$"04 00 00 20 04 00 00 20 04 00 00 20 07 FF FF E0",
		/* [2] */
		$"00 00 00 00 00 00 3F E0 00 00 7F F0 00 00 7F F0"
		$"00 01 FF F0 00 03 FF F0 00 07 FF F0 00 07 FF F0"
		$"00 07 FF F0 00 07 FF F0 00 07 FF F0 00 07 FF F0"
		$"00 0F FF F0 00 0F FF F0 00 1F FF F0 00 3F FF F0"
		$"03 FF FF F0 07 FF FF F0 0F FF FF F0 0F FF FF F0"
		$"0F FF FF F0 0F FF FF F0 0F FF FF F0 0F FF FF F0"
		$"0F FF FF F0 0F FF FF F0 0F FF FF F0 07 FF FF E0"
		$"07 FF FF E0 07 FF FF E0 07 FF FF E0 07 FF FF E0"
	}
};

resource 'DITL' (-4064) {
	{ /* array DITLarray: 1 elements */
		/* [1] */
		{60, 110, 76, 280},
		EditText {
			enabled, ""
		};
		/* [2] */
		{85, 110, 101, 280},
		EditText {
			enabled, ""
		};
		/* [3] */
		{15, 110, 50, 280},
		StaticText {
			disabled, "Apple Macintosh Developer Technical Support TextEdit Control Panel Device Example © 1988"
		}
	}
};

resource 'FREF' (-4064, purgeable) {
	'cdev', 0, ""
};

resource 'nrct' (-4064, purgeable) {
	{	/* array RectArray: 1 elements */
		/* [1] */
		{-1, 87, 130, 322}
	}
};

resource 'mach' (-4064, purgeable) {
	0xFFFF,
	0
};

File: MPW/3.4.2/Examples/CExamples/EditCdev.c
/*------------------------------------------------------------------------------
#
#	Macintosh Developer Technical Support
#
#	EditText Sample Control Panel Device
#
#	EditCdev
#
#	EditCdev.c	-	C Source
#
#	Copyright © 1988, 1995 Apple Computer, Inc.
#	All rights reserved.
#
#	Versions:	1.0					8/88
#
#	Components:	EditCdev.c			August 1, 1988
#				EditCdev.r			August 1, 1988
#				EditCdev.make		August 1, 1988
#
#	EditCdev is a sample Control Panel device (cdev) that 
#	demonstrates the usage of the edit-related messages.  
#	EditCdev demonstrates how to implement an editText item
#	in a Control Panel Device.  It utilizes the new undo, cut, copy,
#	paste, and delete messages that are sent to cdevs in
#	response to user menu selections.
#
#	It is comprised of two editText items that can be edited 
#	and moved between via the mouse or tab key.
#
------------------------------------------------------------------------------*/

#include <Types.h>
#include <Memory.h>
#include <Quickdraw.h>
#include <TextEdit.h>
#include <Dialogs.h>
#include <Events.h>
#include <Devices.h>
#include <Scrap.h>

/* Constants */
#define	textItm		1			/* first editTExt item in cdev */

#define	undoDev		9			/* cdev edit messages */
#define	cutDev		10
#define	copyDev		11
#define	pasteDev	12
#define	clearDev	13


/* Types */
typedef struct CDEVRec {
	TEHandle	myTE;
} CDEVRec, *CDEVPtr, **CDEVHnd;
	
/* Prototypes */
void DoEditCommand (short message, DialogPtr CPDialog);
pascal Handle
TextCDEV(short message, short item, short numItems, short CPanelID,
		 EventRecord *theEvent, Handle cdevStorage, DialogPtr CPDialog);

/* This is the main dispatcher. It must be the first code in the cdev.
 EditCdev's dispatcher responds only to the following messages from
 the Control Panel:
 	
	macDev		- To indicate what machines it is available on.
	initDev		- To set up some temporary storage and get the caret started.
	keyEvtDev	- To check for an edit command and do the appropriate action.
	cutDev		- To cut the current selection.
	copyDev		- To copy the current selection.
	pasteDev	- To paste the contents of the clipboard.
	clearDev	- To delete the current selection.

 The Dialog Manager's services are used to handle entry of text, selection
 of text, editing of text, and moving between the editText items via the
 tab key. Since the Dialog Manager handles the selection of text, we do not
 have to be concerned with hitDev messages for the editText items. The only
 things we have to take care of are calling the Dialog Manager editing
 routines in response to an edit command, and getting the caret to show up
 at the beginning. In response to an edit command that was the result of
 a command-key equivalent, we must also eliminate the event so that it does
 not get processed as a keyDown by the Dialog Manager. Otherwise, an 'x'
 would show up in the editText item when the user did a command-x to cut
 the text.*/
 
pascal Handle
TextCDEV(
	short		message,
	short		item,
	short		numItems,
	short		CPanelID,
	EventRecord	*theEvent,
	Handle		cdevStorage,
	DialogPtr	CPDialog
)
{
	#pragma unused (item, CPanelID)		/* unused formal parameters */
	
	char		tempChar;

	if (message == macDev) return((Handle) 1);				/* we work on every machine */
	else if (cdevStorage != nil) {
		switch (message) {
			case initDev:									/* initialize cdev */
				cdevStorage = NewHandle(sizeof(CDEVRec));	/* create provate storage */
				SelectDialogItemText(CPDialog, numItems + textItm, 0, 999); /* make caret show up */
				break;
				
			case hitDev:									/* handle hit on item */
			case closeDev:									/* clean up and dispose */
			case nulDev:
			case updateDev:									/* handle any update drawing */
			case activDev:									/* activate any needed items */
			case deactivDev:								/* deactivate any needed items */
				break;
				
			case keyEvtDev:									/* respond to keydown */
				tempChar = theEvent->message & charCodeMask;/* get the character, and check */
				if (theEvent->modifiers & cmdKey) {			/*  status of command key */
					message = nulDev;						/* start with no message */
					theEvent->what = nullEvent;				/* and empty event type */
					
					switch (tempChar) {						/* set appropriate message */
						
						case 'X':
						case 'x':
							message = cutDev;
							break;
						case 'C':
						case 'c':
							message = copyDev;
							break;
						case 'V':
						case 'v':
							message = pasteDev;
							break;
					}
					DoEditCommand(message, CPDialog);		/* Let edit command handler take it */
				}
				break;
				
			case macDev:
			case undoDev:
				break;
				
			case cutDev:
			case copyDev:
			case pasteDev:
			case clearDev:
				DoEditCommand(message, CPDialog);			/* respond to edit command */
				break;
		}

		return (cdevStorage);
	}  /* cdevStorage != nil */
	
	/* 
	**	if cdevStorage = NIL then ControlPanel 
	**  will put up memory error
	*/
	return (nil);
}

/* Call the appropriate Dialog Manager routine to handle an edit command for
 	an editText item. It will do all the work regarding the TEScrap. */
void DoEditCommand (short message, DialogPtr CPDialog)
{
	switch (message) {
	case cutDev:
		DialogCut(CPDialog);
		break;
	case copyDev:
		DialogCopy(CPDialog);
		break;
	case pasteDev:
		DialogPaste(CPDialog);
		break;
	case clearDev:
		DialogDelete(CPDialog);
		break;
	}
}

